# YOUR FIRST LAB
### Please read this section. This is valuable to get you prepared, even if it's a long read -- it's important stuff.

### Also, be sure to read [README.md](../README.md)! More info about the updated videos in the README and [top of the course resources in purple](https://edwarddonner.com/2024/11/13/llm-engineering-resources/)

## Your first Frontier LLM Project

By the end of this course, you will have built an autonomous Agentic AI solution with 7 agents that collaborate to solve a business problem. All in good time! We will start with something smaller...

Our goal is to code a new kind of Web Browser. Give it a URL, and it will respond with a summary. The Reader's Digest of the internet!!

Before starting, you should have completed the setup linked in the README.

### If you're new to working in "Notebooks" (also known as Labs or Jupyter Lab)

Welcome to the wonderful world of Data Science experimentation! Simply click in each "cell" with code in it, such as the cell immediately below this text, and hit Shift+Return to execute that cell. Be sure to run every cell, starting at the top, in order.

Please look in the [Guides folder](../guides/01_intro.ipynb) for all the guides.

## I am here to help

If you have any problems at all, please do reach out.  
I'm available through the platform, or at ed@edwarddonner.com, or at https://www.linkedin.com/in/eddonner/ if you'd like to connect (and I love connecting!)   
And I also have a [YouTube channel with extra content](https://youtube.com/@edward.donner).  
And this is new to me, but I'm also trying out X at [@edwarddonner](https://x.com/edwarddonner) - if you're on X, please show me how it's done 😂  

## More troubleshooting

Please see the [troubleshooting](../setup/troubleshooting.ipynb) notebook in the setup folder to diagnose and fix common problems. At the very end of it is a diagnostics script with some useful debug info.

## If this is old hat!

If you're already comfortable with today's material, please hang in there; you can move swiftly through the first few labs - we will get much more in depth as the weeks progress. Ultimately we will fine-tune our own LLM to compete with OpenAI!

<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/important.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#900;">Please read - important note</h2>
            <span style="color:#900;">The way I collaborate with you may be different to other courses you've taken. I prefer not to type code while you watch. Rather, I execute Jupyter Labs, like this, and give you an intuition for what's going on. My suggestion is that you carefully execute this yourself, <b>after</b> watching the lecture. Add print statements to understand what's going on, and then come up with your own variations. If you have a Github account, use this to showcase your variations. Not only is this essential practice, but it demonstrates your skills to others, including perhaps future clients or employers...</span>
        </td>
    </tr>
</table>
<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/resources.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#f71;">This code is a live resource - keep an eye out for my emails</h2>
            <span style="color:#f71;">I push updates to the code regularly. As people ask questions, I add more examples or improved commentary. As a result, you'll notice that the code below isn't identical to the videos. Everything from the videos is here; but I've also added better explanations and new models like DeepSeek. Consider this like an interactive book.<br/><br/>
                I try to send emails regularly with important updates related to the course. You can find this in the 'Announcements' section of Udemy in the left sidebar. You can also choose to receive my emails via your Notification Settings in Udemy. I'm respectful of your inbox and always try to add value with my emails!
            </span>
        </td>
    </tr>
</table>
<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/business.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#181;">Business value of these exercises</h2>
            <span style="color:#181;">A final thought. While I've designed these notebooks to be educational, I've also tried to make them enjoyable. We'll do fun things like have LLMs tell jokes and argue with each other. But fundamentally, my goal is to teach skills you can apply in business. I'll explain business implications as we go, and it's worth keeping this in mind: as you build experience with models and techniques, think of ways you could put this into action at work today. Please do contact me if you'd like to discuss more or if you have ideas to bounce off me.</span>
        </td>
    </tr>
</table>

### If necessary, install Cursor Extensions

1. From the View menu, select Extensions
2. Search for Python
3. Click on "Python" made by "ms-python" and select Install if not already installed
4. Search for Jupyter
5. Click on "Jupyter" made by "ms-toolsai" and select Install if not already installed


### Next Select the Kernel

Click on "Select Kernel" on the Top Right

Choose "Python Environments..."

Then choose the one that looks like `.venv (Python 3.12.x) .venv/bin/python` - it should be marked as "Recommended" and have a big star next to it.

Any problems with this? Head over to the troubleshooting.

### Note: you'll need to set the Kernel with every notebook..

In [1]:
# imports

import os
from dotenv import load_dotenv
from scraper import fetch_website_contents
from IPython.display import Markdown, display
# from openai import OpenAI
from anthropic import Anthropic

# If you get an error running this cell, then please head over to the troubleshooting notebook!

# Connecting to OpenAI (or Ollama)

The next cell is where we load in the environment variables in your `.env` file and connect to OpenAI.  

If you'd like to use free Ollama instead, please see the README section "Free Alternative to Paid APIs", and if you're not sure how to do this, there's a full solution in the solutions folder (day1_with_ollama.ipynb).

## Troubleshooting if you have problems:

If you get a "Name Error" - have you run all cells from the top down? Head over to the Python Foundations guide for a bulletproof way to find and fix all Name Errors.

If that doesn't fix it, head over to the [troubleshooting](../setup/troubleshooting.ipynb) notebook for step by step code to identify the root cause and fix it!

Or, contact me! Message me or email ed@edwarddonner.com and we will get this to work.

Any concerns about API costs? See my notes in the README - costs should be minimal, and you can control it at every point. You can also use Ollama as a free alternative, which we discuss during Day 2.

In [2]:
# Load environment variables in a file called .env

load_dotenv(override=True)
api_key = os.getenv('ANTHROPIC_API_KEY')

# Check the key

if not api_key:
    print("No API key was found - please head over to the troubleshooting notebook in this folder to identify & fix!")
elif not api_key.startswith("sk-proj-"):
    print("An API key was found, but it doesn't start sk-proj-; please check you're using the right key - see troubleshooting notebook")
elif api_key.strip() != api_key:
    print("An API key was found, but it looks like it might have space or tab characters at the start or end - please remove them - see troubleshooting notebook")
else:
    print("API key found and looks good so far!")


An API key was found, but it doesn't start sk-proj-; please check you're using the right key - see troubleshooting notebook


# Let's make a quick call to a Frontier model to get started, as a preview!

In [ ]:
# To give you a preview -- calling OpenAI with these messages is this easy. Any problems, head over to the Troubleshooting notebook.

message = "Hello, Claude! This is my first ever message to you! Hi!"

messages = [{"role": "user", "content": message}]

messages


[{'role': 'user',
  'content': 'Hello, GPT! This is my first ever message to you! Hi!'}]

In [4]:
client = Anthropic(api_key=api_key)

response = client.messages.create(model="claude-haiku-4-5", messages=messages, max_tokens=1024)
response.content[0].text

"Hello! Welcome! It's great to meet you too! 👋\n\nI'm Claude (not GPT, but I appreciate the greeting!). I'm happy to help with whatever you'd like to chat about, ask questions on, work through, or create together.\n\nWhat brings you here today? Anything on your mind you'd like to explore?"

## OK onwards with our first project

In [5]:
# Let's try out this utility

ed = fetch_website_contents("https://edwarddonner.com")
print(ed)

Home - Edward Donner

Skip to content
Avatar
Curriculum
Proficiency
C4
Outsmart
An arena that pits LLMs against each other in a battle of diplomacy and deviousness
About
Posts
Well, hi there.
I’m Ed. I like writing code and experimenting with LLMs, and hopefully you’re here because you do too. I also enjoy amateur electronic music production (
very
amateur) and losing myself in
Hacker News
, nodding my head sagely to things I only half understand.
I’m the co-founder and CTO of AI startup
Nebula.io
. I was previously founder and CEO of AI startup untapt,
acquired in 2021
, and a Managing Director at JPMorgan.
I will happily drone on for hours about LLMs to anyone in my vicinity. My friends got fed up with my impromptu lectures, and convinced me to make some Udemy courses. To my total joy (and shock) they’ve become best-selling, top-rated courses, with 600,000 enrolled across 194 countries. The
full curriculum is here
. If you’re visiting from one of my courses – I’m super grateful!
For 

## Types of prompts

You may know this already - but if not, you will get very familiar with it!

Models like GPT have been trained to receive instructions in a particular way.

They expect to receive:

**A system prompt** that tells them what task they are performing and what tone they should use

**A user prompt** -- the conversation starter that they should reply to

In [6]:
# Define our system prompt - you can experiment with this later, changing the last sentence to 'Respond in markdown in Spanish."

system_prompt = """
You are a snarky assistant that analyzes the contents of a website,
and provides a short, snarky, humorous summary, ignoring text that might be navigation related.
Respond in markdown. Do not wrap the markdown in a code block - respond just with the markdown.
"""

In [7]:
# Define our user prompt

user_prompt_prefix = """
Here are the contents of a website.
Provide a short summary of this website.
If it includes news or announcements, then summarize these too.

"""

## Messages

The API from OpenAI expects to receive messages in a particular structure.
Many of the other APIs share this structure:

```python
[
    {"role": "system", "content": "system message goes here"},
    {"role": "user", "content": "user message goes here"}
]
```
To give you a preview, the next 2 cells make a rather simple call - we won't stretch the mighty GPT (yet!)

In [11]:
messages = [
    {"role": "user", "content": "What is 2 + 2?"}
]

response = client.messages.create(
    model="claude-haiku-4-5",
    system="You are a helpful assistant",
    messages=messages,
    max_tokens=1024
)

response.content[0].text

'2 + 2 = 4'

## And now let's build useful messages for GPT-4.1-mini, using a function

In [12]:
# See how this function creates exactly the format above

def messages_for(website):
    return [
        {"role": "user", "content": user_prompt_prefix + website}
    ]

In [13]:
# Try this out, and then try for a few more websites

messages_for(ed)

[{'role': 'user',
  'content': '\nHere are the contents of a website.\nProvide a short summary of this website.\nIf it includes news or announcements, then summarize these too.\n\nHome - Edward Donner\n\nSkip to content\nAvatar\nCurriculum\nProficiency\nC4\nOutsmart\nAn arena that pits LLMs against each other in a battle of diplomacy and deviousness\nAbout\nPosts\nWell, hi there.\nI’m Ed. I like writing code and experimenting with LLMs, and hopefully you’re here because you do too. I also enjoy amateur electronic music production (\nvery\namateur) and losing myself in\nHacker News\n, nodding my head sagely to things I only half understand.\nI’m the co-founder and CTO of AI startup\nNebula.io\n. I was previously founder and CEO of AI startup untapt,\nacquired in 2021\n, and a Managing Director at JPMorgan.\nI will happily drone on for hours about LLMs to anyone in my vicinity. My friends got fed up with my impromptu lectures, and convinced me to make some Udemy courses. To my total joy 

## Time to bring it together - the API for OpenAI is very simple!

In [16]:
# And now: call the OpenAI API. You will get very familiar with this!

def summarize(url):
    website = fetch_website_contents(url)

    response = client.messages.create(
        model="claude-haiku-4-5",
        messages=messages_for(website),
        system=system_prompt,
        max_tokens=1024
    )

    return response.content[0].text

In [17]:
summarize("https://edwarddonner.com")

'# Edward Donner\'s Website: A Self-Proclaimed LLM Evangelist\n\n**TL;DR:** Ed is a guy who literally cannot stop talking about LLMs (his friends staged an intervention), so he monetized his obsession into best-selling Udemy courses. He\'s now the CTO of an AI startup and apparently has a digital avatar you can chat with if you want to hear about AI without him physically cornering you at parties.\n\n## Key Highlights:\n\n- **The Pitch:** Ed likes code, LLMs, amateur electronic music, and Hacker News (the holy trinity of people who say things like "I\'m not a cryptobro, but...").\n\n- **Credentials:** Co-founder/CTO at Nebula.io, previously ran a startup that got acquired in 2021, and had a fancy JPMorgan title. So he\'s got some legit experience to back up all the LLM preaching.\n\n- **The Success Story:** His unsolicited LLM lectures somehow became 600,000-student best-selling courses across 194 countries. Your friends\' intervention could be someone else\'s goldmine.\n\n- **Recent A

# Edward Donner's Website: A Self-Proclaimed LLM Evangelist

**TL;DR:** Ed is a guy who literally cannot stop talking about LLMs (his friends even staged an intervention), so he turned that obsession into a series of best-selling Udemy courses. Today, he's the CTO of an AI startup and even has a digital avatar you can chat with if you'd like to hear about AI without being cornered at a party.

## Key Highlights

### The Pitch
Ed enjoys:
- Coding
- Large Language Models (LLMs)
- Amateur electronic music
- Hacker News

A combination that might sound familiar if you've ever heard someone begin a sentence with, "I'm not a cryptobro, but..."

### Credentials
- Co-founder and CTO at Nebula.io
- Previously founded a startup that was acquired in 2021
- Former Executive Director at JPMorgan

In other words, there's substantial industry experience behind the AI enthusiasm.

### The Success Story
What started as unsolicited lectures about LLMs somehow evolved into best-selling online courses with more than **600,000 students across 194 countries**.

Apparently, one person's intervention-worthy obsession can become another person's career.

### Recent Announcements

#### February 2026
- New resources for AI Coder tracks
- Agentic Engineering content

#### January 2026
- New materials covering n8n Agents

#### September 2025
- MLOps deployment resources and updates

## Bonus Feature
Ed now has a digital avatar you can chat with.

Because, apparently, regular Ed wasn't enough Ed.

In [18]:
# A function to display this nicely in the output, using markdown

def display_summary(url):
    summary = summarize(url)
    display(Markdown(summary))

In [19]:
display_summary("https://edwarddonner.com")

# Edward Donner's Website Summary

Welcome to the digital home of Ed, a self-proclaimed LLM enthusiast who's made a lucrative hobby out of nerding about artificial intelligence. This site is basically his brain dump, except organized and monetized.

**The Highlights:**
- **Who he is:** Co-founder/CTO at Nebula.io, former startup founder (untapt, acquired 2021), reformed JPMorgan executive, and apparently a very persuasive lecturer
- **His claim to fame:** Created Udemy courses on AI that somehow convinced 600,000 people across 194 countries to pay money to listen to him talk about LLMs
- **His vibe:** Code enthusiast, LLM evangelist, amateur music producer, and HackerNews lurker who finally got friends to convince him to stop cornering people at parties about machine learning

**Recent News/Resources:**
- **Feb 2026:** "AI Coder: Vibe Coder to Agentic Engineer" course resources launched
- **Jan 2026:** AI Builder with n8n course for creating Agents and Voice Agents
- **Sept 2025:** AI Engineering MLOps Track for deploying AI to production
- **May 2025:** Posted course ordering guide (because apparently the curriculum got complicated)

He's also got a digital avatar you can chat with, which is either cutting-edge or deeply unsettling depending on your perspective.

# Let's try more websites

Note that this will only work on websites that can be scraped using this simplistic approach.

Websites that are rendered with Javascript, like React apps, won't show up. See the community-contributions folder for a Selenium implementation that gets around this. You'll need to read up on installing Selenium (ask ChatGPT!)

Also Websites protected with CloudFront (and similar) may give 403 errors - many thanks Andy J for pointing this out.

But many websites will work just fine!

In [20]:
display_summary("https://cnn.com")

# CNN: Breaking News and Feedback Forms

Oh, you've stumbled upon CNN's homepage—a thrilling digital experience dedicated to asking you how you *feel* about ads while simultaneously bombarding you with them. 

**The Vibe:** This is basically a navigation menu's fever dream. The page is absolutely *drowning* in categories (US, World, Politics, Business, Health, Entertainment, Style, Travel, Sports, Science, Climate, Weather, and various global conflicts) because apparently choosing what news to read is like picking from a restaurant with 47 pages.

**The Main Attraction:** About 80% of the visible content is dedicated to a customer feedback form asking whether the video player was slow, the ad was too loud, or if content moved around while loading. Revolutionary journalism at its finest.

**What They're Covering:** 
- World Cup 2026
- Ukraine-Russia War
- Israel-Hamas War
- Plus the usual suspects: elections, Trump, markets, weather

**The Subtext:** CNN really wants to know if you hate their ads. Like, *really* wants to know. They've created a multi-layered feedback system to capture your frustrations in meticulous detail.

*Spoiler alert: You can also follow topics and sign into your CNN account if you enjoy having your news preferences permanently recorded.*

In [21]:
display_summary("https://anthropic.com")

# Anthropic Website Summary

Welcome to Anthropic's home turf—where they sell you on the idea that AI safety is *totally* their top priority (and also their entire business model). 

## The Vibe
This is a public benefit corporation that wants you to know they're *really serious* about making AI safe while simultaneously building increasingly powerful AI systems. The irony is apparently lost on no one, but hey, at least they're transparent about it.

## What They're Hawking
- **Claude**: Their AI assistant that's "just genuinely helpful conversations" (no ads, no sponsored content—just pure, unadulterated AI)
- **Various Claude flavors**: Opus, Sonnet, and Haiku (apparently naming their models after poetry styles adds philosophical gravitas)
- **Claude Code, Claude Cowork, Claude Security, Claude Platform**: Because one Claude wasn't enough, they had to make five of them

## Recent News
- **Claude Opus 4.8** (May 28, 2026): An "upgrade" to Opus that's supposedly better at coding, agentic tasks, and professional work. Translation: *it does more stuff*
- **Project Glasswing**: Securing critical software for the AI era—because nothing says "we care about safety" like a mission-driven project with a mysterious code name

## The Bottom Line
Anthropic wants to be your friendly neighborhood AI safety company while also being a product-driven tech company. Make of that what you will.

<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/business.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#181;">Business applications</h2>
            <span style="color:#181;">In this exercise, you experienced calling the Cloud API of a Frontier Model (a leading model at the frontier of AI) for the first time. We will be using APIs like OpenAI at many stages in the course, in addition to building our own LLMs.

More specifically, we've applied this to Summarization - a classic Gen AI use case to make a summary. This can be applied to any business vertical - summarizing the news, summarizing financial performance, summarizing a resume in a cover letter - the applications are limitless. Consider how you could apply Summarization in your business, and try prototyping a solution.</span>
        </td>
    </tr>
</table>

<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/important.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#900;">Before you continue - now try yourself</h2>
            <span style="color:#900;">Use the cell below to make your own simple commercial example. Stick with the summarization use case for now. Here's an idea: write something that will take the contents of an email, and will suggest an appropriate short subject line for the email. That's the kind of feature that might be built into a commercial email tool.</span>
        </td>
    </tr>
</table>

In [23]:
# Step 1: Create your prompts

system_prompt = "You are a helpful assistant that can read the email text and suggest a catchy subject line for the email."
user_prompt = """
    Here is the email text:
"""
email_text = """
    BookMyShowM-Ticket	BOOKING ID:
WMZWD7P
Hey , your booking is confirmed!
Please find your M-Ticket attached to this email.

Nehru Planetarium, Delhi
Venue  DirectionsDirections
Nehru Planetarium: New Delhi., Delhi-NCR, Delhi
Date & Time
Sun, 7 Jun, 2026| 6:00pm
Category	Seat(s)	Price
ADULT
PLANETAR - D22, D23
₹200.00
View Tickets
Total amount paid	₹223.60
Ticket price	₹200.00
Convenience Fee	₹23.60
Donation to Underprivileged Musicians	₹0.00
Paid using  
How to use M-Ticket:
We have sent M-ticket to you via SMS & Whatsapp also.
M-ticket is available in your purchase history section of your BookMyShow profile.
Show the QR Code present in your M-Ticket to at the entry of event.

"""
# Step 2: Make the messages list

messages = [
    {
        'role': 'user',
        'content': user_prompt + email_text
    }
] # fill this in

# Step 3: Call Claude
response = client.messages.create(
    model="claude-haiku-4-5",
    messages=messages,
    system=system_prompt,
    max_tokens=1024
)

# Step 4: print the result
print(response.content[0].text)

# Suggested Subject Lines:

1. **Your Nehru Planetarium Booking Confirmed - M-Ticket Inside** ✓ (Best for clarity)

2. **Booking Confirmed! Your M-Ticket for Nehru Planetarium (Jun 7)**

3. **Your M-Ticket is Ready - Nehru Planetarium, Delhi**

4. **Confirmation: Planetarium Tickets Booked for June 7, 2026**

5. **M-Ticket Attached - Your Nehru Planetarium Booking**

**My Top Pick:** Option 1 combines the key information (confirmation status, venue, and M-ticket availability) in a clear, scannable format that helps users immediately understand the email's purpose.


## An extra exercise for those who enjoy web scraping

You may notice that if you try `display_summary("https://openai.com")` - it doesn't work! That's because OpenAI has a fancy website that uses Javascript. There are many ways around this that some of you might be familiar with. For example, Selenium is a hugely popular framework that runs a browser behind the scenes, renders the page, and allows you to query it. If you have experience with Selenium, Playwright or similar, then feel free to improve the Website class to use them. In the community-contributions folder, you'll find an example Selenium solution from a student (thank you!)

# Sharing your code

I'd love it if you share your code afterwards so I can share it with others! You'll notice that some students have already made changes (including a Selenium implementation) which you will find in the community-contributions folder. If you'd like add your changes to that folder, submit a Pull Request with your new versions in that folder and I'll merge your changes.

If you're not an expert with git (and I am not!) then I've given you complete instructions in the guides folder, guide 3, and pasting here:

Here's the overall steps involved in making a PR and the key instructions:  
https://edwarddonner.com/pr  

Please check before submitting:  
1. Your PR only contains changes in community-contributions (unless we've discussed it)  
2. All notebook outputs are clear  
3. Less than 2,000 lines of code in total, and not too many files  
4. Don't include unnecessary test files, or overly wordy README or .env.example or emojis or other LLM artifacts!

Thanks so much!

Detailed steps here: 

https://chatgpt.com/share/6873c22b-2a1c-8012-bc9a-debdcf7c835b